### PR TITLE
[SYCL-MLIR] Add target extension type

### PR DESCRIPTION
**Pull request by sommerlukas**
Add support for the `llvm::TargetExtType` to the MLIR LLVM dialect. 

Target extension types are used to represent target-specific types, which are opaque to the compiler, when using opaque pointers. 

They are relevant for us, as they are used to represent SPIR-V types such as images.

Also enforces some of the constraints defined for target extension types by the LLVM language reference manual, e.g., only `zeroinitializer` constants.

### CODE REVIEW

**Comment by sommerlukas** [comment]
I intend to also push this as a patch to upstream LLVM once reviewed and approved here.

While this PR adds the support for the type itself, it does not yet introduce the correct handling in the `cgeist` codegen, to fix the tests affected by the lack of this type so far. This will be added in a separate PR.

**Review by whitneywhtsang — APPROVED** [review]

File: mlir/lib/Dialect/LLVMIR/IR/LLVMTypes.cpp [hunk 1]
```diff
@@ -109,6 +109,56 @@ static void printPointer(AsmPrinter &p, Type elementType,
   }
 }
 
+//===----------------------------------------------------------------------===//
+// custom<ExtTypeParams>
+//===----------------------------------------------------------------------===//
+
+static bool parseTypeOrIntParam(AsmParser &p, SmallVector<Type> &typeParams,
```
[the remark this answers is not in the document]
Comment (sommerlukas) [reply mlir/lib/Dialect/LLVMIR/IR/LLVMTypes.cpp:116]: If you prefer, I can change this, I just tried to adopt the style used for other parsers in this file, see for example `parseFunctionTypes` above, which all use `SmallVector&` as a function parameter.

[the remark this answers is not in the document]
Comment (Naghasan) [reply mlir/lib/Dialect/LLVMIR/IR/LLVMTypes.cpp:116]: IIRC it is a best practice as it avoids to link to a specific stack length. Probably best to do the change regardless.

[the remark this answers is not in the document]
Comment (sommerlukas) [reply mlir/lib/Dialect/LLVMIR/IR/LLVMTypes.cpp]: Ok, done.

**Review by etiotto — APPROVED** [review]

**Review by Naghasan — APPROVED** [review]

**Comment by sommerlukas** [comment]
Upstream patch: https://reviews.llvm.org/D151446

**Comment by sommerlukas** [comment]
The upstream patch has been accepted and merged. 
I updated this PR with the changes from the upstream patch, so they should now be identical.

**Pull request merged**